src/unix/stream.c

```diff
@@ -886,6 +891,34 @@ static void uv__write(uv_stream_t* stream) {
     /* Ensure the handle isn't sent again in case this is a partial write. */
     if (n >= 0)
       req->send_handle = NULL;
+#if defined(TCP_FASTOPEN) && defined(MSG_FASTOPEN)
+  } else if (stream->flags & UV_HANDLE_TCP_FASTOPEN) {
+    struct sockaddr_storage* addr;
+    socklen_t addrlen;
+
+    addr = (struct sockaddr_storage*)&((uv_tcp_t*)stream)->addr;
+    switch (addr->ss_family) {
+      case AF_INET:
+        addrlen = sizeof(struct sockaddr_in);
+        break;
+      case AF_INET6:
+        addrlen = sizeof(struct sockaddr_in6);
+        break;
+      default:
+        err = UV__ERR(EINVAL);
+        goto error;
+    }
+    do {
+      n = sendto(uv__stream_fd(stream), iov[0].iov_base, iov[0].iov_len,
+                 MSG_FASTOPEN, (const struct sockaddr*)addr, addrlen);
+    } while (n == -1 && (RETRY_ON_WRITE_ERROR(errno) || errno == EINPROGRESS));
```

Comment:
The `EINPROGRESS` has to be handled differently than re-trying `sendto()` again (which only results in `EALREADY`).

I think when `EINPROGRESS` is returned, it should be handled similarly to when it happens in `uv__tcp_connect()`. The code as is doesn't seem to be able to handle subsequent writes to the stream until the connection is established.